text/0000-core_detect.md

```diff
@@ -0,0 +1,91 @@
+
+A new set of macros is introduced, named ``core`::arch::mark_*_feature_as_detected`, which accept the same feature names as the equivalent `is_*_feature_detected` macros. These macros are unsafe to call: the caller must ensure that the relevant CPU feature is actually available.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Static initialization
+
+Runtime feature detection is performed in `std` using a static initializer, which allows it work both when built as a program and as a static/dynamic library. This would perform the same work that is currently done by `std_detect` when the target features are first queried.
+
+## Thread-safety
+
+Modifying the available target features must be thread-safe. This is achieved by treating the target features as an array of `AtomicUsize` which are filled in with `AtomicUsize::fetch_or`.
+
+For platforms without support for compare-exchange, this is instead represented as an array of `AtomicBool`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Possible breaking change
+
+The effect of `mark_*_feature_as_detected` is one-way: it is only possible to enable a feature, not disable them. However this does introduce a minor breaking change: the value returned by `is_*_feature_detected` could change from `false` to `true` at any point. This could potentially cause issues if some code in another thread is depending on `is_*_feature_detected` to select a code path, and switching code paths halfway through would produce an incorrect result.
+
+There are 2 ways in which we can address this:
+- Declare that it's users' responsibility to ensure their code is robust against CPU features changing at runtime.
+- Require that `mark_*_feature_as_detected` is only called when no other threads are concurrently running. However this may be difficult/impossible when Rust code is compiled as a dynamic library: it could be loaded while other threads are already running.
+
+## Use of static initializers
+
+The use of static initializers is somewhat controversial, for example see https://github.com/rust-lang/rust/issues/111921.
+
+## Size of `core`
+
+This increases the size of the `.bss` section of `libcore` by a few bytes, which can have an impact on very minimial binaries.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Using a lang item to call back into `std`
+
+Instead of having `std` "push" the CPU features to `core` at initialization time, an alternative design would be for `core` to "pull" this information from `std` by calling a lang item defined in `std`. The problem with this approach is that it doesn't provide a clear path for how this would be exposed to no-std programs which want to do their own feature detection.
```

Comment:
A lang item with two "grades" - weak and strong, core defines a weak version of the lang item so users are not *required* to provide their own lang item definition, but anyone can override it with a strong version of the same lang item (libstd will do that as well).

Reply:
(A lang item can have a stable alias like `#[panic_handler]`.)

Reply:
Weak lang items is already used as name for lang items like `#[panic_handler]` that need not be defined when it is used, but does need to be defined when linking if it was used anywhere. Maybe use preemptible lang items as name here?